common/scanner.h

```diff
@@ -1,294 +1,444 @@
+    if ((vec).cap < (_cap)) {                                                  \
+        STRING_RESIZE((vec), (_cap));                                          \
+    }
 
-  unsigned serialize(char *buffer) {
-    size_t size = quoted_string_id.size();
+#define STRING_PUSH(vec, el)                                                   \
+    if ((vec).cap == (vec).len) {                                              \
+        STRING_RESIZE((vec), MAX(16, (vec).len * 2));                          \
+    }                                                                          \
+    (vec).data[(vec).len++] = (el);
 
-    buffer[0] = in_string;
-    quoted_string_id.copy(&buffer[1], size);
-    return size + 1;
-  }
+#define STRING_FREE(vec)                                                       \
+    if ((vec).data != NULL)                                                    \
+        free((vec).data);
 
-  void deserialize(const char *buffer, unsigned length) {
-    if (length > 0) {
-      in_string = buffer[0];
-      quoted_string_id.assign(&buffer[1], length - 1);
-    }
-  }
-
-  void advance(TSLexer *lexer) {
-    lexer->advance(lexer, false);
-  }
-
-  void skip(TSLexer *lexer) {
-    lexer->advance(lexer, true);
-  }
-
-  bool scan(TSLexer *lexer, const bool *valid_symbols) {
-    if (valid_symbols[LEFT_QUOTED_STRING_DELIM] && (iswlower(lexer->lookahead) || lexer->lookahead == '_' || lexer->lookahead == '|')) {
-      lexer->result_symbol = LEFT_QUOTED_STRING_DELIM;
-      return scan_left_quoted_string_delimiter(lexer);
-    } else if (valid_symbols[RIGHT_QUOTED_STRING_DELIM] && (lexer->lookahead == '|')) {
-      advance(lexer);
-      lexer->result_symbol = RIGHT_QUOTED_STRING_DELIM;
-      return scan_right_quoted_string_delimiter(lexer);
-    } else if (in_string && valid_symbols[STRING_DELIM] && lexer->lookahead == '"') {
-      advance(lexer);
-      in_string = false;
-      lexer->result_symbol = STRING_DELIM;
-      return true;
+#define STRING_CLEAR(vec)                                                      \
+    {                                                                          \
+        (vec).len = 0;                                                         \
+        memset((vec).data, 0, (vec).cap * sizeof(char));                       \
     }
 
-    while (iswspace(lexer->lookahead)) {
-      skip(lexer);
-    }
+enum TokenType {
+    COMMENT,
+    LEFT_QUOTED_STRING_DELIM,
+    RIGHT_QUOTED_STRING_DELIM,
+    STRING_DELIM,
+    LINE_NUMBER_DIRECTIVE,
+    NULL_CHARACTER
+};
 
-    if (!in_string && lexer->lookahead == '#' && lexer->get_column(lexer) == 0) {
-      advance(lexer);
+typedef struct {
+    uint32_t cap;
+    uint32_t len;
+    char *data;
+} String;
 
-      while (lexer->lookahead == ' ' || lexer->lookahead == '\t') advance(lexer);
+String string_new() {
+    return (String){.cap = 16, .len = 0, .data = calloc(1, sizeof(char) * 17)};
```

Comment:
Should we `assert` that `calloc()` doesn't return `NULL`? If you do it in `STRING_RESIZE()`, it's weird not to do it here.

Reply:
hmm yeah, that's a good idea